### ISSUE TITLE
HTML5 project export no longer runs in 3.0 beta2.

### ISSUE THREAD

**Issue by planetxpress** [godotengine/godot#15015]
**Godot version:**
3.0 beta2

**OS/device including version:**
Exported on Windows 10 Pro v1709. Attempted to run on latest versions of Google Chrome and Mozilla Firefox.

**Issue description:**
HTML5 exported project (with updated beta2 export templates) fails to run in browser with 2 errors.
```
**ERROR**: exit(0) implicitly called by end of main(), but noExitRuntime, so not exiting the runtime (you can use emscripten_force_exit, if you want to force a true shutdown)
'ccall' was not exported. add it to EXTRA_EXPORTED_RUNTIME_METHODS (see the FAQ)
```

Saw this environment variable in [platform/javascript/detect.py](https://github.com/godotengine/godot/blob/master/platform/javascript/detect.py#L108) however I  suspect I do not have the requisite knowledge to re-compile the app and templates properly, or understand what other dependencies this might have.

**Steps to reproduce:**
Load a project and export project to HTML5 with default settings.

**Minimal reproduction project:**
[Pong](https://github.com/godotengine/godot-demo-projects/tree/master/2d/pong) in the Godot demo projects, for example.

- [X] I searched the existing [GitHub issues](https://github.com/godotengine/godot/issues?utf8=%E2%9C%93&q=is%3Aissue+) for potential duplicates.

**Comment by joaopedrosgs** [comment]
same here

**Comment by goldenratio** [comment]
full error log from browser console

```
still waiting on run dependencies:
game.html:366:6
dependency: wasm-instantiate
game.html:366:6
(end of list)
game.html:366:6
pre-main prep time: 17383 ms
game.html:366:6
let it go dude!
game.html:361:6
Calling stub instead of sigaction()
game.html:366:6
warning: build with  -s DEMANGLE_SUPPORT=1  to link in libcxxabi demangling
game.html:366:6
Init OS
game.html:361:6
Init Audio
game.html:361:6
Init VS
game.html:361:6
Init Physicsserver
game.html:361:6
OpenGL ES 3.0 Renderer: Mozilla
game.html:361:6
GLES3: max ubo light: 409
game.html:361:6
GLES3: max ubo reflections: 455, ubo size: 144
game.html:361:6
4096
game.js:1:36341
ARVR: Registered interface: Native mobile
game.html:361:6
exit(0) implicitly called by end of main(), but noExitRuntime, so not exiting the runtime (you can use emscripten_force_exit, if you want to force a true shutdown)
game.html:366:6
'ccall' was not exported. add it to EXTRA_EXPORTED_RUNTIME_METHODS (see the FAQ)
game.html:361:6
'ccall' was not exported. add it to EXTRA_EXPORTED_RUNTIME_METHODS (see the FAQ)
game.html:366:6
uncaught exception: abort("'ccall' was not exported. add it to EXTRA_EXPORTED_RUNTIME_METHODS (see the FAQ)") at jsStackTrace@http://localhost:8080/godot3_wasm/game.js:1:18605
stackTrace@http://localhost:8080/godot3_wasm/game.js:1:18776
abort@http://localhost:8080/godot3_wasm/game.js:1:456782
RuntimeEnvironment/Module.ccall@http://localhost:8080/godot3_wasm/game.js:1:9643
RuntimeEnvironment/ASM_CONSTS</<@http://localhost:8080/godot3_wasm/game.js:1:37053
doCallback@http://localhost:8080/godot3_wasm/game.js:1:80039
done@http://localhost:8080/godot3_wasm/game.js:1:80177
reconcile@http://localhost:8080/godot3_wasm/game.js:1:62835
RuntimeEnvironment/syncfs/</<@http://localhost:8080/godot3_wasm/game.js:1:58796
@http://localhost:8080/godot3_wasm/game.js:1:60631

```

**Comment by haxpor** [comment]
Same here. I linked the project with C++ custom module, and saw the same error output. The project is just test, print a message, and call custom module's method. Other platforms are ok.

**Comment by guilhermehto** [comment]
Same problem here.

Had the exact same error in both Godot 3.0 Beta1 and 2

Windows 8.1 / Firefox 57.0.2 (64-bit)

Log:

```
pre-main prep time: 9263 ms
betav3.html:366:6
let it go dude!
betav3.html:361:6
Calling stub instead of sigaction()
betav3.html:366:6
warning: build with  -s DEMANGLE_SUPPORT=1  to link in libcxxabi demangling
betav3.html:366:6
Init OS
betav3.html:361:6
Init Audio
betav3.html:361:6
Init VS
betav3.html:361:6
Init Physicsserver
betav3.html:361:6
OpenGL ES 3.0 Renderer: Mozilla
betav3.html:361:6
GLES3: max ubo light: 409
betav3.html:361:6
GLES3: max ubo reflections: 455, ubo size: 144
betav3.html:361:6
4096
betav3.js:1:36341
ARVR: Registered interface: Native mobile
betav3.html:361:6
exit(0) implicitly called by end of main(), but noExitRuntime, so not exiting the runtime (you can use emscripten_force_exit, if you want to force a true shutdown)
betav3.html:366:6
'ccall' was not exported. add it to EXTRA_EXPORTED_RUNTIME_METHODS (see the FAQ)
betav3.html:361:6
'ccall' was not exported. add it to EXTRA_EXPORTED_RUNTIME_METHODS (see the FAQ)
betav3.html:366:6
uncaught exception: abort("'ccall' was not exported. add it to EXTRA_EXPORTED_RUNTIME_METHODS (see the FAQ)") at jsStackTrace@file:///C:/Godot/Builds/Betav3html5/betav3.js:1:18605
stackTrace@file:///C:/Godot/Builds/Betav3html5/betav3.js:1:18776
abort@file:///C:/Godot/Builds/Betav3html5/betav3.js:1:456782
RuntimeEnvironment/Module.ccall@file:///C:/Godot/Builds/Betav3html5/betav3.js:1:9643
RuntimeEnvironment/ASM_CONSTS</<@file:///C:/Godot/Builds/Betav3html5/betav3.js:1:37053
doCallback@file:///C:/Godot/Builds/Betav3html5/betav3.js:1:80039
done@file:///C:/Godot/Builds/Betav3html5/betav3.js:1:80177
reconcile@file:///C:/Godot/Builds/Betav3html5/betav3.js:1:62835
RuntimeEnvironment/syncfs/</<@file:///C:/Godot/Builds/Betav3html5/betav3.js:1:58796
@file:///C:/Godot/Builds/Betav3html5/betav3.js:1:60631
```

**Comment by Acvarium** [comment]
Same problem.

**Comment by sudoio** [comment]
Was trying to fix this. Changed this line:
https://github.com/godotengine/godot/blob/003f2dab786c76ff0e9041b173ea6c7b3b572403/platform/javascript/detect.py#L108
To:
```py
env.Append(LINKFLAGS=['-s', 'EXTRA_EXPORTED_RUNTIME_METHODS="[\'FS\',\'ccall\']"'])
```
After recompiling (#15365 should be merged in order for compilation to work), there were similar issues with other functions, so I had to add those to that line too:

```py
env.Append(LINKFLAGS=['-s', 'EXTRA_EXPORTED_RUNTIME_METHODS="[\'FS\',\'ccall\',\'stringToUTF8\',\'UTF8ToString\',\'cwrap\']"'])
```

However, after these fixes, there is a different issue:
```
**ERROR**: pre-main prep time: 488 ms
let it go dude!
**ERROR**: Calling stub instead of sigaction()
**ERROR**: warning: build with  -s DEMANGLE_SUPPORT=1  to link in libcxxabi demangling
Init OS
Init Audio
Init VS
Init Physicsserver
OpenGL ES 3.0 Renderer: Mozilla
GLES3: max ubo light: 409
GLES3: max ubo reflections: 455, ubo size: 144
ARVR: Registered interface: Native mobile
**ERROR**: enlarged memory arrays from 16777216 to 33554432, took 0 ms (has ArrayBuffer.transfer? false)
```

Should I open a PR for this partial fix?

**Comment by KoltesDigital** [comment]
It may be useful to see which commit broke the export, it may be something else as just adding `ccall` to the exported symbols.

**Comment by jedStevens** [comment]
Do we know what is causing the `**ERROR**: enlarged memory arrays from 16777216 to 33554432, took 1 ms (has ArrayBuffer.transfer? false)` error? I'm just following along.

**Comment by KoltesDigital** [comment]
With this new version, my html5 build loads further. However there's another error when compiling the vertex shader *TonemapShaderGLES3*.

I don't have the error about resizing arrays, maybe because my error occurs first.

**Comment by jedStevens** [comment]
I guess it go either way, maybe you aren't including something I did?

Either way I think it ought to get its own issue since its blocking the export process, I just don't know what to label it as; since things seem to split from here.

**Comment by eska014** [comment]
@jedStevens The WebAssembly memory buffer grows as needed, this is expected and not an issue. Emscripten reports this to stderr, so it's says `**ERROR**`

@KoltesDigital Can't reproduce, try with #15443, otherwise please open an issue with details

**Comment by jedStevens** [comment]
@eska014 this is the last message I got before a black screen HTML5 export, it would seem like this is the culprit no?

**Comment by eska014** [comment]
No, black screen is already fixed in #15443

**Comment by KoltesDigital** [comment]
@eska014 It doesn't help. #15441 opened.

Thanks for this fix.

**Comment by josx** [comment]
@sudoio  @jedStevens 
I am using 3.0.2stable and i am trying to export the pong example to html5.

I am geting this:

```
**ERROR**: still waiting on run dependencies:
**ERROR**: dependency: wasm-instantiate
**ERROR**: (end of list)
let it go dude!
**ERROR**: Calling stub instead of sigaction()
**ERROR**: warning: build with  -s DEMANGLE_SUPPORT=1  to link in libcxxabi demangling
Init OS
Init Audio
Init VS
Init Physicsserver
OpenGL ES 3.0 Renderer: WebKit WebGL
**ERROR**: enlarged memory arrays from 16777216 to 33554432, took 73 ms (has ArrayBuffer.transfer? false)
```

Is there any fix or workaround?

**Comment by j2l** [comment]
Also getting it using Godot 3.1beta3 on an empty project exported to html5

**Comment by wyrmling** [comment]
> Also getting it using Godot 3.1beta3 on an empty project exported to html5, on Chrome and FF:

Did you try to open it on local server? Without server, opening html directly, game will probably not start.

**Comment by j2l** [comment]
I always publish my tests (as temp github project with github pages, or as glitch.com project).
